[LV] Improve accuracy of branch weights in epilogue iteration check block

david-arm opened the pull request:
When one of the vector loops (main or epilogue) is scalable and the
other isn't, we can use the estimated value of vscale to improve the
accuracy.

david-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -7649,6 +7649,16 @@ EpilogueVectorizerEpilogueLoop::emitMinimumVectorEpilogueIterCountCheck(
         EPI.EpilogueUF * EPI.EpilogueVF.getKnownMinValue();
+    // When one of the vector loops is scalable and the other isn't, we can use
+    // the estimated value of vscale to improve the accuracy.
+    if (EPI.MainLoopVF.isScalable() != EPI.EpilogueVF.isScalable()) {
```
The `EPI.MainLoopVF.isScalable() != EPI.EpilogueVF.isScalable()` check guarantees that only one of them is scalable. For the case when they're both scalable the weights are correct and I was trying to avoid doing unnecessary work. I could just always multiply by vscale and call `estimateElementCount` instead if you prefer?

lukel97 replied:
Yeah I was thinking that might be a bit simpler since there's less cases to consider, and that means the steps are always in concrete amounts

lukel97 reviewed (approved):
LGTM

david-arm committed:
Add tests

david-arm committed:
[LV] Improve accuracy of branch weights in epilogue iteration check block

When one of the vector loops (main or epilogue) is scalable and the
other isn't, we can use the estimated value of vscale to improve the
accuracy.

david-arm committed:
Address review comment

fhahn commented on llvm/test/Transforms/LoopVectorize/AArch64/check-prof-info.ll:
```diff
@@ -10,15 +12,15 @@ target triple = "aarch64-unknown-linux-gnu"
 ; vscale=2 for neoverse-v1 and vscale=1 for neoverse-v2.
-define void @_Z3foov(i64 %n) {
-; CHECK-V1-IC1-LABEL: define void @_Z3foov(
+define void @foo_i32(i64 %n) vscale_range(1,16) {
```
Sounds good, thanks
